etcdctl/ctlv3/command/watch_command.go

```diff
@@ -125,14 +157,54 @@ func getWatchChan(c *clientv3.Client, args []string) (clientv3.WatchChan, error)
 	if watchPrevKey {
 		opts = append(opts, clientv3.WithPrevKV())
 	}
-	return c.Watch(clientv3.WithRequireLeader(context.Background()), key, opts...), nil
+	return c.Watch(clientv3.WithRequireLeader(context.Background()), key, opts...), execArgs, nil
 }
 
-func printWatchCh(ch clientv3.WatchChan) {
+func printWatchCh(c *clientv3.Client, ch clientv3.WatchChan, execArgs []string) {
 	for resp := range ch {
 		if resp.Canceled {
 			fmt.Fprintf(os.Stderr, "watch was canceled (%v)\n", resp.Err())
 		}
 		display.Watch(resp)
+
+		if len(execArgs) > 0 {
+			cmd := exec.CommandContext(c.Ctx(), execArgs[0], execArgs[1:]...)
+			cmd.Env = os.Environ()
+			cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
+			if err := cmd.Run(); err != nil {
+				fmt.Fprintf(os.Stderr, "command %q error (%v)\n", execArgs, err)
+			}
+		}
+	}
+}
+
+func findExecArgs(args []string) ([]string, error) {
+	foundSep, idx := false, 0
+	for idx = range args {
+		if args[idx] == "--" && idx != 0 {
+			foundSep = true
+			break
+		}
+	}
+	if !foundSep {
+		return nil, nil
+	}
+
+	if idx == len(args)-1 {
```

Comment:
Let's split the args and execArgs into two distinct lists here and then simplify all following, logic, e.g.:

```
func findExecArgs(args []string) ([]string, error) {

foundSep, idx := false, 0
 for idx = range args {
 	if args[idx] == "--" && idx != 0 {
 		foundSep = true
 		break
 	}
 }
 if !foundSep {
 	return args, nil, nil
 }
watchArgs, execArgs := args[:idx], args[idx+1:]

// check for various invalid states
...

return watchArgs, execArgs, nil

}
```